d2/src/main/java/com/linkedin/d2/xds/XdsClientImpl.java

```diff
@@ -0,0 +1,709 @@
+      if (subscriber == null)
+      {
+        _log.info("Subscribe LDS resource {}", resourceName);
+        subscriber = new ResourceSubscriber(ResourceType.LDS, resourceName);
+        _ldsResourceSubscribers.put(resourceName, subscriber);
+
+        if (_adsStream == null)
+        {
+          startRpcStream();
+        }
+        _adsStream.sendDiscoveryRequest(ResourceType.LDS, Collections.singletonList(resourceName));
+      }
+      subscriber.addWatcher(watcher);
+    });
+  }
+
+  @Override
+  void watchRdsResource(final String resourceName, final RdsResourceWatcher watcher)
+  {
+    _executorService.execute(() ->
+    {
+      ResourceSubscriber subscriber = _rdsResourceSubscribers.get(resourceName);
+      if (subscriber == null)
+      {
+        _log.info("Subscribe RDS resource {}", resourceName);
+        subscriber = new ResourceSubscriber(ResourceType.RDS, resourceName);
+        _rdsResourceSubscribers.put(resourceName, subscriber);
+
+        if (_adsStream == null)
+        {
+          startRpcStream();
+        }
+        _adsStream.sendDiscoveryRequest(ResourceType.RDS, Collections.singletonList(resourceName));
+      }
+      subscriber.addWatcher(watcher);
+    });
+  }
+
+  @Override
+  void watchCdsResource(final String resourceName, final CdsResourceWatcher watcher)
+  {
+    _executorService.execute(() ->
+    {
+      ResourceSubscriber subscriber = _cdsResourceSubscribers.get(resourceName);
+      if (subscriber == null)
+      {
+        _log.info("Subscribe CDS resource {}", resourceName);
+        subscriber = new ResourceSubscriber(ResourceType.CDS, resourceName);
+        _cdsResourceSubscribers.put(resourceName, subscriber);
+
+        if (_adsStream == null)
+        {
+          startRpcStream();
+        }
+        _adsStream.sendDiscoveryRequest(ResourceType.CDS, Collections.singletonList(resourceName));
+      }
+      subscriber.addWatcher(watcher);
+    });
+  }
+
+  @Override
+  void watchEdsResource(final String resourceName, final EdsResourceWatcher watcher)
+  {
+    _executorService.execute(() ->
+    {
+      ResourceSubscriber subscriber = _edsResourceSubscribers.get(resourceName);
+      if (subscriber == null)
+      {
+        _log.info("Subscribe EDS resource {}", resourceName);
+        subscriber = new ResourceSubscriber(ResourceType.EDS, resourceName);
+        _edsResourceSubscribers.put(resourceName, subscriber);
+
+        if (_adsStream == null)
+        {
+          startRpcStream();
+        }
+        _adsStream.sendDiscoveryRequest(ResourceType.EDS, Collections.singletonList(resourceName));
+      }
+      subscriber.addWatcher(watcher);
+    });
+  }
+
+  private void handleLdsResponse(DiscoveryResponseData responseData)
+  {
+    Map<String, LdsUpdate> ldsUpdates = new HashMap<>();
+    List<String> errors = new ArrayList<>();
+
+    for (Resource resource : responseData.getResourcesList())
+    {
+      try
```

Comment:
Consider adding debug traces for the received resource name and version, these are super useful in the future in debugging xDS server-client interactions.

Reply:
Added